File: output/RBTreeFFICheatcode.php
<?php
$exports['runRBTreeFFICheatcode'] = function($limit) {
    $n = (int)$limit;
    $root = null;
    
    $insert = function($root, $key, $value) use (&$insert) {
        if ($root === null) {
            return (object)["color" => 'B', "left" => null, "key" => $key, "value" => $value, "right" => null];
        }
        
        $balance = function($color, $left, $key, $value, $right) {
            if ($color === 'B') {
                if ($left !== null && $left->color === 'R') {
                    if ($left->left !== null && $left->left->color === 'R') {
                        return (object)[
                            "color" => 'R',
                            "left" => (object)["color" => 'B', "left" => $left->left->left, "key" => $left->left->key, "value" => $left->left->value, "right" => $left->left->right],
                            "key" => $left->key,
                            "value" => $left->value,
                            "right" => (object)["color" => 'B', "left" => $left->right, "key" => $key, "value" => $value, "right" => $right]
                        ];
                    }
                    if ($left->right !== null && $left->right->color === 'R') {
                        return (object)[
                            "color" => 'R',
                            "left" => (object)["color" => 'B', "left" => $left->left, "key" => $left->key, "value" => $left->value, "right" => $left->right->left],
                            "key" => $left->right->key,
                            "value" => $left->right->value,
                            "right" => (object)["color" => 'B', "left" => $left->right->right, "key" => $key, "value" => $value, "right" => $right]
                        ];
                    }
                }
                if ($right !== null && $right->color === 'R') {
                    if ($right->left !== null && $right->left->color === 'R') {
                        return (object)[
                            "color" => 'R',
                            "left" => (object)["color" => 'B', "left" => $left, "key" => $key, "value" => $value, "right" => $right->left->left],
                            "key" => $right->left->key,
                            "value" => $right->left->value,
                            "right" => (object)["color" => 'B', "left" => $right->left->right, "key" => $right->key, "value" => $right->value, "right" => $right->right]
                        ];
                    }
                    if ($right->right !== null && $right->right->color === 'R') {
                        return (object)[
                            "color" => 'R',
                            "left" => (object)["color" => 'B', "left" => $left, "key" => $key, "value" => $value, "right" => $right->left],
                            "key" => $right->key,
                            "value" => $right->value,
                            "right" => (object)["color" => 'B', "left" => $right->right->left, "key" => $right->right->key, "value" => $right->right->value, "right" => $right->right->right]
                        ];
                    }
                }
            }
            return (object)["color" => $color, "left" => $left, "key" => $key, "value" => $value, "right" => $right];
        };

        $ins = function($node) use (&$ins, $key, $value, $balance) {
            if ($node === null) {
                return (object)["color" => 'R', "left" => null, "key" => $key, "value" => $value, "right" => null];
            }
            if ($key < $node->key) {
                return $balance($node->color, $ins($node->left), $node->key, $node->value, $node->right);
            } else if ($key > $node->key) {
                return $balance($node->color, $node->left, $node->key, $node->value, $ins($node->right));
            } else {
                return (object)["color" => $node->color, "left" => $node->left, "key" => $key, "value" => $value, "right" => $node->right];
            }
        };
        
        $res = $ins($root);
        $res->color = 'B';
        return $res;
    };
    
    for ($i = 0; $i < $n; $i++) {
        $root = $insert($root, $i, $i);
    }
    
    $lookup = function($node, $key) {
        while ($node !== null) {
            if ($key < $node->key) {
                $node = $node->left;
            } else if ($key > $node->key) {
                $node = $node->right;
            } else {
                return $node;
            }
        }
        return null;
    };

    $sum = 0;
    for ($i = 0; $i < $n; $i++) {
        $node = $lookup($root, $i);
        if ($node !== null) {
            $sum += $node->value;
        }
    }
    return $sum;
};
return $exports;

File: scratch/gopurs-adt-reuse-validation-20260916/after-fficc/src/Test/RBTreeFFICheatcode.cc
#include "purescript.h"
#include <algorithm>
#include <memory>
#include <utility>

namespace {
enum Color { Red, Black };
struct Node {
    Color color;
    int key;
    std::unique_ptr<Node> left;
    std::unique_ptr<Node> right;
};
using Tree = std::unique_ptr<Node>;

bool red(const Tree& tree) { return tree && tree->color == Red; }

Tree rotate_left(Tree tree) {
    Tree root = std::move(tree->right);
    tree->right = std::move(root->left);
    root->left = std::move(tree);
    return root;
}

Tree rotate_right(Tree tree) {
    Tree root = std::move(tree->left);
    tree->left = std::move(root->right);
    root->right = std::move(tree);
    return root;
}

Tree recolor(Tree tree) {
    tree->color = Red;
    tree->left->color = Black;
    tree->right->color = Black;
    return tree;
}

Tree balance(Tree tree) {
    if (tree->color != Black) return tree;
    // The four Okasaki cases, in the same order as Test.RBTree.balance.
    if (red(tree->left) && red(tree->left->left)) {
        return recolor(rotate_right(std::move(tree)));
    }
    if (red(tree->left) && red(tree->left->right)) {
        tree->left = rotate_left(std::move(tree->left));
        return recolor(rotate_right(std::move(tree)));
    }
    if (red(tree->right) && red(tree->right->left)) {
        tree->right = rotate_right(std::move(tree->right));
        return recolor(rotate_left(std::move(tree)));
    }
    if (red(tree->right) && red(tree->right->right)) {
        return recolor(rotate_left(std::move(tree)));
    }
    return tree;
}

Tree ins(Tree tree, int key) {
    if (!tree) return Tree(new Node{Red, key, nullptr, nullptr});
    if (key < tree->key) tree->left = ins(std::move(tree->left), key);
    else if (key > tree->key) tree->right = ins(std::move(tree->right), key);
    else return tree;
    return balance(std::move(tree));
}

Tree insert(Tree tree, int key) {
    tree = ins(std::move(tree), key);
    tree->color = Black;
    return tree;
}

int depth(const Tree& tree) {
    if (!tree) return 0;
    return 1 + std::max(depth(tree->left), depth(tree->right));
}
}

FOREIGN_BEGIN( Test_RBTreeFFICheatcode )
exports["runRBTreeFFICheatcode"] = [](const boxed& in) -> boxed {
    Tree tree;
    for (int n = unbox<int>(in); n > 0; --n) tree = insert(std::move(tree), n);
    return depth(tree);
};
FOREIGN_END
